chunkset_tpl.h

```diff
@@ -112,66 +66,85 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
        Assert(len >= sizeof(uint64_t), "chunkmemset should be called on larger chunks"); */
     Assert(dist > 0, "chunkmemset cannot have a distance 0");
 
-    unsigned char *from = out - dist;
-    chunk_t chunk;
-    unsigned sz = sizeof(chunk);
-    if (len < sz) {
-        while (len != 0) {
-            *out++ = *from++;
-            --len;
-        }
-        return out;
-    }
+    uint8_t *from = out-dist;
 
-#ifdef HAVE_CHUNKMEMSET_1
     if (dist == 1) {
-        chunkmemset_1(from, &chunk);
-    } else
-#endif
+        memset(out, *from, len);
+        return out + len;
+    } else if (dist > sizeof(chunk_t)) {
+        out = CHUNKUNROLL(out, &dist, &len);
+        return CHUNKCOPY(out, out - dist, len);
+    }
+
+    chunk_t chunk_load;
+
+    /* TODO: possibily build up a permutation table for this if not an even
+     * modulus */
 #ifdef HAVE_CHUNKMEMSET_2
     if (dist == 2) {
-        chunkmemset_2(from, &chunk);
+        chunkmemset_2(from, &chunk_load);
     } else
 #endif
 #ifdef HAVE_CHUNKMEMSET_4
     if (dist == 4) {
-        chunkmemset_4(from, &chunk);
+        chunkmemset_4(from, &chunk_load);
     } else
 #endif
 #ifdef HAVE_CHUNKMEMSET_8
     if (dist == 8) {
-        chunkmemset_8(from, &chunk);
+        chunkmemset_8(from, &chunk_load);
+    } else if (dist == sizeof(chunk_t)) {
+        loadchunk(from, &chunk_load);
     } else
 #endif
-    if (dist == sz) {
-        loadchunk(from, &chunk);
-    } else if (dist < sz) {
-        unsigned char *end = out + len - 1;
-        while (len > dist) {
-            out = CHUNKCOPY_SAFE(out, from, dist, end);
-            len -= dist;
+    {
+        int cpy_dist;
+        int bytes_remaining = sizeof(chunk_t);
+        uint8_t *cur_chunk = (uint8_t*)&chunk_load;
+        while (bytes_remaining) {
+            cpy_dist = MIN(dist, bytes_remaining);
+            memcpy(cur_chunk, from, cpy_dist);
+            bytes_remaining -= cpy_dist;
+            cur_chunk += cpy_dist;
         }
-        if (len > 0) {
-            out = CHUNKCOPY_SAFE(out, from, len, end);
+    }
+
+    /* if we're lucky enough and dist happens to be an
+     * even modulus of our vector length, we can do two
+     * stores per loop iteration, which for most ISAs,
+     * especially x86, is beneficial */
+    unsigned chunk_mod = sizeof(chunk_t) % dist;
+    if (chunk_mod == 0) {
+        while (len >= (2*sizeof(chunk_t))) {
+            storechunk(out, &chunk_load);
+            storechunk(out+sizeof(chunk_t), &chunk_load);
+            out += 2*sizeof(chunk_t);
+            len -= 2*sizeof(chunk_t);
+        }
+
+        while (len >= sizeof(chunk_t)) {
+            storechunk(out, &chunk_load);
+            out += sizeof(chunk_t);
+            len -= sizeof(chunk_t);
         }
-        return out;
     } else {
-        out = CHUNKUNROLL(out, &dist, &len);
-        return CHUNKCOPY(out, out - dist, len);
+        /* This division is slow but the compiler, at least on
```

Comment:
Not sure if this comment is necessary since we don't actually do a division right?

Reply:
We do, earlier (implicitly with the modulus with a non-constant).  It's still slow but the line following that comment no longer divides as I found it was also doing a (relatively) expensive imul instruction instead of just doing a subtraction.  Using the subtraction here resulted in some fairly good savings.